presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -556,42 +582,186 @@ private static NestedField nestedField(String name)
+                    }
                     return null;
                 }
+                if (isPushDownSubfieldsFromLambdasEnabled) {
+                    Optional<Set<Subfield>> lambdaSubfieldsOriginal = context.getLambdaSubfields();
+                    if ((functionDescriptor.isAccessingInputValues() && functionDescriptor.getLambdaDescriptors().isEmpty())
+                            || (functionDescriptor.isTransformingInput()) && !functionDescriptor.getOutputToInputTransformationFunction().isPresent()) {
+                        // If function internally accesses the input values we cannot prune any unaccessed subfields since we do not know what subfields function accessed.
+                        // If function transforms the shape of the input map or array argument and the form of such transformation was not supplied in
+                        // getOutputToInputTransformationFunction() then we have to discard all collected accessed subfields in lambdas of outer calls because we do not know how
+                        // to attribute those subfields to the input argument.
+                        // Additionally, we have to stop collecting lambda subfields in both cases until we find another function that does not return input values.
+                        context.setLambdaSubfields(Optional.empty());
+                    }
+                    if (!functionDescriptor.isReturningInputValues()) {
+                        context.setLambdaSubfields(Optional.of(ImmutableSet.of()));
+                    }
 
-                // visit subscript expressions only
-                Optional<Subfield> subfield = toSubfield(call, functionResolution, expressionOptimizer, connectorSession);
-                if (subfield.isPresent()) {
-                    context.subfields.add(subfield.get());
+                    transformOuterLambdaSubfieldsIfNeeded(context, functionDescriptor);
+
+                    if (context.getLambdaSubfields().isPresent()) {
+                        Set<Integer> argumentIndicesContainingMapOrArray = functionDescriptor.getArgumentIndicesContainingMapOrArray()
+                                .orElse(IntStream.range(0, call.getArguments().size()).boxed().collect(Collectors.toSet()));
+                        Map<Integer, Set<Subfield>> argumentIndexToLambdaSubfieldsMap = argumentIndicesContainingMapOrArray.stream()
+                                .collect(Collectors.toMap(callArgumentIndex -> callArgumentIndex, unused -> new HashSet<>(context.getLambdaSubfields().get())));
+
+                        if (functionDescriptor.isAccessingInputValues() && !functionDescriptor.getLambdaDescriptors().isEmpty()) {
+                            for (LambdaDescriptor lambdaDescriptor : functionDescriptor.getLambdaDescriptors()) {
+                                if (!addLambdaSubfields(call, context, lambdaDescriptor, argumentIndexToLambdaSubfieldsMap)) {
```

Comment:
what about the case where a function doesn't return input (e.g. any match), but the lambda doesn't actually access any subfields?  we should add a none subfield then too

Reply:
Added support for this case